api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,616 @@
+		}
+	}
+
+	return shortestDistribution
+}
+
+func (e *EnvironmentScaler) getContainerResourceFromDeploy(deployID string) ([]scaler.ResourceConsumer, error) {
+	if consumers, ok := e.deployCache.Getf(deployID); ok {
+		return consumers.([]scaler.ResourceConsumer), nil
+	}
+
+	taskDefinition, err := e.getTaskDefinitionFromDeployID(deployID)
+	if err != nil {
+		return nil, err
+	}
+
+	consumers := make([]scaler.ResourceConsumer, len(taskDefinition.ContainerDefinitions))
+	for i, taskDefinition := range taskDefinition.ContainerDefinitions {
+		var cpu int
+		var memory bytesize.Bytesize
+
+		if c := int(aws.Int64Value(taskDefinition.Cpu)); c != 0 {
+			cpu = c
+		}
+
+		if m := aws.Int64Value(taskDefinition.MemoryReservation); m != 0 {
+			memory = bytesize.MiB * bytesize.Bytesize(m)
+		}
+
+		if m := aws.Int64Value(taskDefinition.Memory); m != 0 {
+			memory = bytesize.MiB * bytesize.Bytesize(m)
+		}
+
+		ports := []int{}
+		for _, p := range taskDefinition.PortMappings {
+			if hostPort := int(aws.Int64Value(p.HostPort)); hostPort != 0 {
+				ports = append(ports, hostPort)
+			}
+		}
+
+		id := "NewConsumerFromDeploy::" + deployID
+		consumers[i] = scaler.NewResourceConsumer(cpu, id, memory, ports)
+	}
+
+	e.deployCache.Add(deployID, consumers)
+	return consumers, nil
+}
+
+func (e *EnvironmentScaler) getServiceResourceConsumers(clusterName string) ([]scaler.ResourceConsumer, error) {
+	var resourceConsumers []scaler.ResourceConsumer
+
+	serviceARNs, err := e.getServiceARNsForCluster(clusterName)
+	if err != nil {
+		return nil, err
+	}
+
+	services, err := e.getServicesFromServiceARNs(clusterName, serviceARNs)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, service := range services {
+		deployIDCopies := map[string]int{}
+		// deployment.RunningCount is the number of containers already running on an instance
+		// deployment.PendingCount is the number of containers that are alraedy on an instance, but are being pulled
+		// we only care about containers that are not on instances yet
+		for _, d := range service.Deployments {
+			desiredCount := aws.Int64Value(d.DesiredCount)
+			runningCount := aws.Int64Value(d.RunningCount)
+			pendingCount := aws.Int64Value(d.PendingCount)
+			if numPending := desiredCount - (runningCount + pendingCount); numPending > 0 {
+				deployIDCopies[aws.StringValue(d.TaskDefinition)] = int(numPending)
+			}
+		}
+
+		if len(deployIDCopies) == 0 {
+			continue
+		}
+
+		// iterate through deploys and their copies
+		for deployID, numCopies := range deployIDCopies {
+			for i := 0; i < numCopies; i++ {
+				consumers, err := e.getContainerResourceFromDeploy(deployID)
+				if err != nil {
+					return nil, err
+				}
+
+				resourceConsumers = append(resourceConsumers, consumers...)
+			}
+		}
+	}
+
+	return resourceConsumers, nil
+}
+
+func (e *EnvironmentScaler) getJobTaskResourceConsumers(clusterName string) ([]scaler.ResourceConsumer, error) {
+	var resourceConsumers []scaler.ResourceConsumer
```

Comment:
don't use `var foo []Bar` notation for slices, instead use `foo := []Bar{}`. Otherwise, `append` will cause a panic because the variable is nil.